tests/std/tests/P0896R4_counted_iterator/test.cpp

```diff
@@ -0,0 +1,185 @@
+                { // increment assign
+                    auto iter = counted_iterator<Iter>{Iter{input}, 5};
+                    auto ref  = iter;
+                    iter += 3;
+                    assert(iter.count() == 2);
+                    assert(*iter == 4);
+                    assert(iter.count() == ref.count() - 3);
+                    assert(iter.base().peek() == next(ref.base().peek(), 3));
+                }
+                { // decrement by n
+                    auto iter = counted_iterator<Iter>{Iter{begin(input) + 2}, 3};
+                    auto ref  = iter - 2;
+                    assert(iter.count() == 3);
+                    assert(*iter == 3);
+                    assert(iter.count() == ref.count() - 2);
+                    assert(iter.base().peek() == next(ref.base().peek(), 2));
+                }
+            }
+            { // difference
+                auto iter1 = counted_iterator<Iter>{Iter{input + 1}, 2};
+                auto iter2 = counted_iterator<Iter>{Iter{input}, 3};
+
+                const same_as<iter_difference_t<Iter>> auto diff1 = iter1 - iter2;
+                assert(diff1 == 1);
+
+                const same_as<iter_difference_t<Iter>> auto diff2 = iter2 - iter1;
+                assert(diff2 == -1);
+            }
+            // TODO different type that satisfies common_with
+            { // difference default sentinel
+                auto iter1 = counted_iterator<Iter>{Iter{input}, 2};
+
+                const same_as<iter_difference_t<Iter>> auto diff1 = iter1 - default_sentinel;
+                assert(diff1 == -2);
+
+                const same_as<iter_difference_t<Iter>> auto diff2 = default_sentinel - iter1;
+                assert(diff2 == 2);
+            }
+            if constexpr (random_access_iterator<Iter>) { // decrement assign
+                auto iter = counted_iterator<Iter>{Iter{end(input)}, 0};
+                auto ref  = iter;
+                iter -= 3;
+                assert(iter.count() == 3);
+                assert(*iter == 3);
+                assert(iter.count() == ref.count() + 3);
+                assert(iter.base().peek() == prev(ref.base().peek(), 3));
+            }
+        }
+        // [counted.iter.cmp]
+        {
+            { // equality
+                auto iter1 = counted_iterator<Iter>{Iter{input}, 2};
+                auto iter2 = counted_iterator<Iter>{Iter{input}, 2};
+                assert(iter1 == iter2);
+            }
+            { // inequality
+                auto iter1 = counted_iterator<Iter>{Iter{input + 1}, 2};
+                auto iter2 = counted_iterator<Iter>{Iter{input}, 3};
+                assert(iter1 != iter2);
+            }
+            { // equality default sentinel
+                auto iter1 = counted_iterator<Iter>{Iter{input}, 2};
+                auto iter2 = counted_iterator<Iter>{Iter{input}, 0};
+                assert(iter1 != default_sentinel);
+                assert(iter2 == default_sentinel);

```

Comment:
Should we have symmetric test cases to make sure the compiler doesn't fumble the expression rewrites?

Reply:
Added transposed tests